osf/management/commands/collect_user_nodes_exceeding_storage_limits.py

```diff
@@ -0,0 +1,82 @@
+from django.core.management.base import BaseCommand
+import json
+import logging
+
+from api.caching.tasks import update_storage_usage_cache
+from osf.models import Node
+from website.settings import StorageLimits
+
+logger = logging.getLogger(__name__)
+
+
+def get_write_admin_contributors(node):
+    write_admin_contributor_set = set()
+    for contributor in node.contributor_set.all():
+        if contributor.permission in ['write', 'admin']:
+            write_admin_contributor_set.add(contributor.user._id)
+    return write_admin_contributor_set
+
+
+def retrieve_user_nodes_exceeding_storage_limits():
+    exceeded_user_node_dict = dict()
+
+    for node in Node.objects.filter(type='osf.node'):
```

Comment:
The `type` filtering isn't doing anything; `Node.objects` already does so because the manager is for an already-typed model (as opposed to `AbstractNode.objects`). 

`.all()` would work, but our targets here are really only nodes with `OsfStorageFile` objects. An `OuterRef`/`Exists` annotation is much more expensive at the start, but would also cut the number of projects to traverse by about 70%:
```
In [1]: Node.objects.count()
Out[1]: 804251

In [2]: files = OsfStorageFile.objects.filter(target_object_id=OuterRef('pk'), target_content_type_id=30)

In [3]: Node.objects.annotate(has_storage=Exists(files)).filter(has_storage=True).count()
Out[3]: 241644
```
Couple that with `tqdm` and this loop might look something like
```
files = OsfStorageFile.objects.filter(target_object_id=OuterRef('pk'), target_content_type_id=30)
qs = Node.objects.annotate(has_storage=Exists(files)).filter(has_storage=True)
p_bar = tqdm(total=qs.count())

for i, node in enumerate(qs, 1):
    # do the things
    p_bar.update(i)
p_bar.close()
```

Reply:
Really cool - I wasn't familiar with the OuterRef/Exists annotation pattern so thank you for that!

Reply:
Yep. It's not as efficient as constructing a raw SQL query[1], but is much easier to write, and less likely to break with ORM changes.

[1] Gets compiled to something like:
```sql
SELECT
  osf_abstractnode.[model_fields, ...],
  EXISTS([files_subselect]) AS "has_storage"
FROM osf_abstractnode
WHERE EXISTS([files_subselect]) = TRUE
```
so it's executing that subselect twice when only once is strictly necessary (in this case, at least). 
Generally, I'd not recommend it for request-blocking app code without thorough benchmarking.